// @devexpress/utils/lib/utils/base64.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Base64Utils = exports.DocmMimeType = exports.PlainTextMimeType = exports.RtfMimeType = exports.OpenXmlMimeType = exports.Base64MimeType = void 0;
var Base64MimeType;
(function (Base64MimeType) {
    Base64MimeType[Base64MimeType["Unknown"] = 0] = "Unknown";
    Base64MimeType[Base64MimeType["OpenXml"] = 1] = "OpenXml";
    Base64MimeType[Base64MimeType["Rtf"] = 2] = "Rtf";
    Base64MimeType[Base64MimeType["PlainText"] = 3] = "PlainText";
    Base64MimeType[Base64MimeType["Docm"] = 4] = "Docm";
})(Base64MimeType = exports.Base64MimeType || (exports.Base64MimeType = {}));
exports.OpenXmlMimeType = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
exports.RtfMimeType = 'application/rtf';
exports.PlainTextMimeType = 'text/plain';
exports.DocmMimeType = 'application/vnd.ms-word.document.macroEnabled.12';
var Base64Utils = (function () {
    function Base64Utils() {
    }
    Base64Utils.normalizeToDataUrl = function (base64, mimeType) {
        if (!Base64Utils.checkPrependDataUrl(base64))
            base64 = Base64Utils.prependByDataUrl(base64, mimeType);
        return base64;
    };
    Base64Utils.prependByDataUrl = function (base64, mimeType) {
        return "data:" + mimeType + ";base64," + base64;
    };
    Base64Utils.checkPrependDataUrl = function (base64) {
        return Base64Utils.dataUrl.test(base64);
    };
    Base64Utils.deleteDataUrlPrefix = function (base64DataUrl) {
        return base64DataUrl.replace(Base64Utils.dataUrl, '');
    };
    Base64Utils.getUint8Array = function (base64) {
        base64 = atob(base64);
        var n = base64.length;
        var arr = new Uint8Array(n);
        while (n--)
            arr[n] = base64.charCodeAt(n);
        return arr;
    };
    Base64Utils.fromArrayBuffer = function (buffer) {
        var binary = [];
        var bytes = new Uint8Array(buffer);
        var len = bytes.byteLength;
        for (var i = 0; i < len; i++)
            binary.push(String.fromCharCode(bytes[i]));
        return window.btoa(binary.join(''));
    };
    Base64Utils.getFileFromBase64 = function (base64, blobOptions) {
        if (blobOptions === void 0) { blobOptions = { type: 'text' }; }
        var data = Base64Utils.getUint8Array(base64);
        return (new Blob([data], blobOptions));
    };
    Base64Utils.getMimeTypeAsString = function (base64) {
        var match = base64.match(Base64Utils.dataUrl);
        return match ? match[1] : null;
    };
    Base64Utils.getKnownMimeType = function (base64) {
        var _a;
        var match = base64.match(Base64Utils.dataUrl);
        return match ? ((_a = Base64Utils.mimeTypesMap[match[1]]) !== null && _a !== void 0 ? _a : Base64MimeType.Unknown) : Base64MimeType.Unknown;
    };
    Base64Utils.fromBlobAsArrayBuffer = function (blob, callback) {
        var reader = new FileReader();
        reader.onloadend = function () { return callback(Base64Utils.fromArrayBuffer(reader.result)); };
        reader.readAsArrayBuffer(blob);
    };
    Base64Utils.fromBlobAsDataUrl = function (blob, callback) {
        var reader = new FileReader();
        reader.onloadend = function () { return callback(reader.result); };
        reader.readAsDataURL(blob);
    };
    Base64Utils.mimeTypesMap = (_a = {},
        _a[exports.OpenXmlMimeType] = Base64MimeType.OpenXml,
        _a[exports.RtfMimeType] = Base64MimeType.Rtf,
        _a[exports.PlainTextMimeType] = Base64MimeType.PlainText,
        _a[exports.DocmMimeType] = Base64MimeType.Docm,
        _a);
    Base64Utils.dataUrl = /^data:(.*?)(;(.*?))??(;base64)?,/;
    return Base64Utils;
}());
exports.Base64Utils = Base64Utils;
