// @devexpress/dx-react-chart-material-ui/dist/dx-react-chart-material-ui.cjs.js
/**
 * Bundle of @devexpress/dx-react-chart-material-ui
 * Generated: 2020-11-03
 * Version: 2.7.3
 * License: https://js.devexpress.com/Licensing
 */

'use strict';

if (typeof process === "undefined") { var process = { env: {} }; }

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var PropTypes = require('prop-types');
var dxReactCore = require('@devexpress/dx-react-core');
var dxReactChart = require('@devexpress/dx-react-chart');
var colorManipulator = require('@material-ui/core/styles/colorManipulator');
var styles$h = require('@material-ui/core/styles');
var classNames = _interopDefault(require('clsx'));
var List = _interopDefault(require('@material-ui/core/List'));
var ListItemText = _interopDefault(require('@material-ui/core/ListItemText'));
var ListItem = _interopDefault(require('@material-ui/core/ListItem'));
var Typography = _interopDefault(require('@material-ui/core/Typography'));
var Popper = _interopDefault(require('@material-ui/core/Popper'));
var dxChartCore = require('@devexpress/dx-chart-core');
var Paper = _interopDefault(require('@material-ui/core/Paper'));

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) {
    return;
  }

  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

var getBorderColor = function getBorderColor(theme) {
  return theme.palette.type === 'light' ? colorManipulator.lighten(colorManipulator.fade(theme.palette.divider, 1), 0.88) : colorManipulator.darken(colorManipulator.fade(theme.palette.divider, 1), 0.68);
};
var withClassName = function withClassName() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  var setClassName = function setClassName(_ref) {
    var className = _ref.className,
        classes = _ref.classes,
        restProps = _objectWithoutProperties(_ref, ["className", "classes"]);

    return _objectSpread2({
      className: classNames(classes.root, className)
    }, restProps);
  }; // TODO: First candidate to `compose` util?


  return function (Target) {
    return styles$h.withStyles.apply(void 0, args)(dxReactChart.withPatchedProps(setClassName)(Target));
  };
};

var styles = function styles(theme) {
  var _theme$typography = theme.typography,
      fontFamily = _theme$typography.fontFamily,
      fontSize = _theme$typography.fontSize,
      fontWeightLight = _theme$typography.fontWeightLight;
  return {
    root: {
      fontFamily: fontFamily,
      fontSize: fontSize,
      fontWeight: fontWeightLight,
      display: 'flex',
      flexDirection: 'column',
      padding: '10px',
      boxSizing: 'border-box'
    }
  };
};

var Root = withClassName(styles)(dxReactChart.Chart.Root);

var styles$1 = function styles(theme) {
  var fontFamily = theme.typography.fontFamily;
  return {
    root: {
      fill: theme.palette.text.secondary,
      fontFamily: fontFamily,
      fontSize: 12,
      fontWeight: 400
    }
  };
};

var Label = withClassName(styles$1)(dxReactChart.Chart.Label);

var palette = ['#42A5F5', '#FF7043', '#9CCC65', '#FFCA28', '#26A69A', '#EC407A'];

var ChartWithPalette = function ChartWithPalette(_ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"]);

  return React.createElement(dxReactChart.Chart, props, React.createElement(dxReactChart.Palette, {
    scheme: palette
  }), children);
};

ChartWithPalette.components = dxReactChart.Chart.components;
process.env.NODE_ENV !== "production" ? ChartWithPalette.propTypes = {
  children: PropTypes.node.isRequired
} : void 0;
var Chart = dxReactCore.withComponents({
  Root: Root
})(ChartWithPalette);
Chart.Label = Label;

var styles$2 = function styles() {
  return {
    root: {
      paddingTop: 16,
      paddingBottom: 16
    }
  };
};

var Root$1 = withClassName(styles$2, {
  name: 'LegendRoot'
})(List);

var styles$3 = function styles() {
  return {
    root: {
      fontSize: 14,
      paddingLeft: 8,
      paddingRight: 8
    }
  };
};

var Label$1 = withClassName(styles$3, {
  name: 'LegendLabel'
})(function (_ref) {
  var text = _ref.text,
      restProps = _objectWithoutProperties(_ref, ["text"]);

  return React.createElement(ListItemText, restProps, text);
});

var styles$4 = function styles(theme) {
  return {
    root: {
      alignItems: 'center',
      paddingTop: theme.spacing(0.5),
      paddingBottom: theme.spacing(0.5)
    }
  };
};

var Item = withClassName(styles$4, {
  name: 'LegendItem'
})(ListItem);

var Legend = dxReactCore.withComponents({
  Root: Root$1,
  Item: Item,
  Label: Label$1
})(dxReactChart.Legend);

var styles$5 = function styles() {
  return {
    root: {
      textAlign: 'center',
      width: '100%',
      marginBottom: '20px'
    }
  };
};

var Text = withClassName(styles$5, {
  name: 'Title'
})(function (_ref) {
  var text = _ref.text,
      restProps = _objectWithoutProperties(_ref, ["text"]);

  return React.createElement(Typography, _extends({
    component: "h3",
    variant: "h5"
  }, restProps), text);
});

var Title = dxReactCore.withComponents({
  Text: Text
})(dxReactChart.Title);

var styles$6 = function styles(theme) {
  return {
    point: {
      fill: theme.palette.background.paper
    }
  };
};

var setClassName = function setClassName(_ref) {
  var classes = _ref.classes,
      restProps = _objectWithoutProperties(_ref, ["classes"]);

  if (restProps.state) {
    var className = restProps.className,
        rest = _objectWithoutProperties(restProps, ["className"]);

    return _objectSpread2({
      className: classNames(classes.point, className)
    }, rest);
  }

  return restProps;
};

var Point = styles$h.withStyles(styles$6)(dxReactChart.withPatchedProps(setClassName)(dxReactChart.ScatterSeries.Point));

var ScatterSeries = dxReactCore.withComponents({
  Point: Point
})(dxReactChart.ScatterSeries);

var styles$7 = function styles() {
  return {
    root: {
      shapeRendering: 'crispEdges'
    }
  };
};

var Root$2 = withClassName(styles$7)(dxReactChart.Axis.Root);

var styles$8 = function styles(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Tick = withClassName(styles$8)(dxReactChart.Axis.Tick);

var styles$9 = function styles(theme) {
  var fontFamily = theme.typography.fontFamily;
  return {
    root: {
      fill: theme.palette.text.secondary,
      fontFamily: fontFamily,
      fontSize: 12,
      fontWeight: 400,
      backgroundColor: theme.palette.background.paper
    }
  };
};

var Label$2 = withClassName(styles$9)(dxReactChart.Axis.Label);

var styles$a = function styles(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Line = withClassName(styles$a)(dxReactChart.Axis.Line);

var styles$b = function styles(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Grid = withClassName(styles$b)(dxReactChart.Axis.Grid);

var ValueAxis = dxReactCore.withComponents({
  Root: Root$2,
  Tick: Tick,
  Label: Label$2,
  Line: Line,
  Grid: Grid
})(dxReactChart.ValueAxis);

var ArgumentAxis = dxReactCore.withComponents({
  Root: Root$2,
  Tick: Tick,
  Label: Label$2,
  Line: Line,
  Grid: Grid
})(dxReactChart.ArgumentAxis);

var styles$c = function styles(theme) {
  var arrowSize = theme.spacing(1.2);
  return {
    'popper-top': {
      zIndex: 1,
      marginBottom: "".concat(arrowSize, "px")
    },
    'popper-right': {
      zIndex: 1,
      marginLeft: "".concat(arrowSize, "px")
    }
  };
};

var popperModifiers = function popperModifiers(arrowRef) {
  return {
    flip: {
      enabled: false
    },
    arrow: {
      element: arrowRef
    }
  };
};

var OverlayBase = function OverlayBase(_ref) {
  var classes = _ref.classes,
      className = _ref.className,
      children = _ref.children,
      target = _ref.target,
      rotated = _ref.rotated,
      ArrowComponent = _ref.arrowComponent,
      restProps = _objectWithoutProperties(_ref, ["classes", "className", "children", "target", "rotated", "arrowComponent"]);

  var _React$useState = React.useState(null),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      arrowRef = _React$useState2[0],
      setArrowRef = _React$useState2[1];

  var placement = rotated ? dxChartCore.RIGHT : dxChartCore.TOP;
  return React.createElement(Popper, _extends({
    open: true,
    anchorEl: target,
    placement: placement,
    className: classNames(classes["popper-".concat(placement)], className),
    modifiers: popperModifiers(arrowRef)
  }, restProps), children, React.createElement(ArrowComponent, {
    placement: placement,
    ref: setArrowRef
  }));
};

process.env.NODE_ENV !== "production" ? OverlayBase.propTypes = {
  className: PropTypes.string,
  classes: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired,
  target: PropTypes.any.isRequired,
  rotated: PropTypes.bool.isRequired,
  arrowComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]).isRequired
} : void 0;
OverlayBase.defaultProps = {
  className: undefined
};
var Overlay = styles$h.withStyles(styles$c)(OverlayBase);

var styles$d = function styles(theme) {
  return {
    root: {
      fontSize: 14,
      padding: theme.spacing(0.5)
    }
  };
};

var Content = withClassName(styles$d, {
  name: 'TooltipContent'
})(function (_ref) {
  var text = _ref.text,
      targetItem = _ref.targetItem,
      restProps = _objectWithoutProperties(_ref, ["text", "targetItem"]);

  return React.createElement(Typography, restProps, text);
});

var styles$e = function styles(theme) {
  var arrowSize = theme.spacing(1.2);
  return {
    'arrow-top': {
      width: "".concat(arrowSize * 2, "px"),
      height: "".concat(arrowSize, "px"),
      position: 'absolute',
      top: '100%',
      overflow: 'hidden',
      '&::after': {
        content: '""',
        position: 'absolute',
        width: "".concat(arrowSize, "px"),
        height: "".concat(arrowSize, "px"),
        background: theme.palette.background.paper,
        transform: 'translateX(-50%) translateY(-50%) rotate(45deg)',
        left: '50%',
        boxShadow: theme.shadows[2]
      }
    },
    'arrow-right': {
      width: "".concat(arrowSize, "px"),
      height: "".concat(arrowSize * 2, "px"),
      position: 'absolute',
      top: '50%',
      transform: 'translateX(-100%)',
      overflow: 'hidden',
      '&::after': {
        content: '""',
        position: 'absolute',
        width: "".concat(arrowSize, "px"),
        height: "".concat(arrowSize, "px"),
        background: theme.palette.background.paper,
        transform: 'translateX(-50%) translateY(-50%) rotate(45deg)',
        top: '50%',
        left: '100%',
        boxShadow: theme.shadows[2]
      }
    }
  };
};

var BaseArrow = React.forwardRef(function (_ref, ref) {
  var classes = _ref.classes,
      className = _ref.className,
      placement = _ref.placement,
      restProps = _objectWithoutProperties(_ref, ["classes", "className", "placement"]);

  return React.createElement("div", _extends({
    className: classNames(classes["arrow-".concat(placement)], className),
    ref: ref
  }, restProps));
});
BaseArrow.propTypes = {
  placement: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  className: PropTypes.string
};
BaseArrow.defaultProps = {
  className: undefined
};
var Arrow = styles$h.withStyles(styles$e)(BaseArrow);

var styles$f = function styles(theme) {
  return {
    root: {
      padding: theme.spacing(0.5, 1)
    }
  };
};

var Sheet = withClassName(styles$f)(function (props) {
  return React.createElement(Paper, props);
});

var Tooltip = dxReactCore.withComponents({
  Overlay: Overlay,
  Content: Content,
  Sheet: Sheet,
  Arrow: Arrow
})(dxReactChart.Tooltip);

var styles$g = function styles(theme) {
  return {
    root: {
      fill: theme.palette.primary.main,
      opacity: 0.2
    }
  };
};

var DragBox = withClassName(styles$g)(dxReactChart.ZoomAndPan.DragBox);

var ZoomAndPan = dxReactCore.withComponents({
  DragBox: DragBox
})(dxReactChart.ZoomAndPan);

Object.defineProperty(exports, 'AreaSeries', {
  enumerable: true,
  get: function () {
    return dxReactChart.AreaSeries;
  }
});
Object.defineProperty(exports, 'BarSeries', {
  enumerable: true,
  get: function () {
    return dxReactChart.BarSeries;
  }
});
Object.defineProperty(exports, 'LineSeries', {
  enumerable: true,
  get: function () {
    return dxReactChart.LineSeries;
  }
});
Object.defineProperty(exports, 'PieSeries', {
  enumerable: true,
  get: function () {
    return dxReactChart.PieSeries;
  }
});
Object.defineProperty(exports, 'SplineSeries', {
  enumerable: true,
  get: function () {
    return dxReactChart.SplineSeries;
  }
});
exports.ArgumentAxis = ArgumentAxis;
exports.Chart = Chart;
exports.Legend = Legend;
exports.ScatterSeries = ScatterSeries;
exports.Title = Title;
exports.Tooltip = Tooltip;
exports.ValueAxis = ValueAxis;
exports.ZoomAndPan = ZoomAndPan;
